File: medium/15. 3Sum/Solution.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        #sort
        #target = -num[i]
        #from num[low] = num[i+1] ~ num[hi] = num[len-1]
        #if num[low] + num[high] < -num[i], low+
        #else >, high -
        result = []
        nums.sort()
        if not nums or nums[-1]< 0 or nums[0] > 0:
            return result
        else:
            for i in range(len(nums)-2):
                if i > 0 and nums[i] == nums[i-1]:
                    continue
                low = i +1
                high = len(nums) -1
                target = -1 * nums[i]
                while low < high:
                    sum = nums[low] + nums[high]
                    if sum < target or (low > i + 1 and nums[low] == nums[low - 1]):
                        low +=1
                    elif sum > target or (high < len(nums) -1 and nums[high]== nums[high+1]):
                        high -=1
                    else:
                        result.append([nums[i],nums[low],nums[high]])
                        low +=1
                        high -=1
        return result

File: medium/15. 3Sum/Solution.java
class Solution {
    
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        Arrays.sort(nums);
        int l = nums.length;
        for (int i = 0 ; i < l-2 ; i++){
            if((i==0)||(nums[i]!=nums[i-1] && i>0)){
            int sum;
            sum = nums[i]*(-1);
            int low = i+1;
            int high = l-1;
            while(low<high){
            int twoSum = 0;
            twoSum = nums[low] + nums[high];
            if(sum == twoSum){
                result.add(Arrays.asList(nums[i],nums[low],nums[high]));
                low++;
                high--;
                while(nums[high]==nums[high+1] && high>0) high--;
                while(nums[low]==nums[low-1] && low < l-1) low++;
            }
            else if(sum < twoSum){
                high--;  
                while(nums[high]==nums[high+1] && high>0) high--;
            }
            else{
                low++;    
                while(nums[low]==nums[low-1] && low < l-1) low++;
            }
            
            }
        }
        }
